src/log-store/src/kafka/client_manager.rs

```diff
@@ -0,0 +1,130 @@
+
+use common_config::wal::{KafkaConfig, KafkaWalTopic as Topic};
+use dashmap::mapref::entry::Entry as DashMapEntry;
+use dashmap::DashMap;
+use rskafka::client::partition::{PartitionClient, UnknownTopicHandling};
+use rskafka::client::producer::aggregator::RecordAggregator;
+use rskafka::client::producer::{BatchProducer, BatchProducerBuilder};
+use rskafka::client::{Client as RsKafkaClient, ClientBuilder};
+use rskafka::BackoffConfig;
+use snafu::ResultExt;
+
+use crate::error::{BuildClientSnafu, BuildPartitionClientSnafu, Result};
+
+// Each topic only has one partition for now.
+// The `DEFAULT_PARTITION` refers to the index of the partition.
+const DEFAULT_PARTITION: i32 = 0;
+
+/// Arc wrapper of Client.
+pub(crate) type ClientRef = Arc<Client>;
+/// Arc wrapper of ClientManager.
+pub(crate) type ClientManagerRef = Arc<ClientManager>;
+
+/// A client through which to contact Kafka cluster. Each client associates with one partition of a topic.
+/// Since a topic only has one partition in our design, the mapping between clients and topics are one-one.
+#[derive(Debug)]
+pub(crate) struct Client {
+    /// A raw client used to construct a batch producer and/or a stream consumer for a specific topic.
+    pub(crate) raw_client: Arc<PartitionClient>,
+    /// A producer used to buffer log entries for a specific topic before sending them in a batching manner.
+    pub(crate) producer: Arc<BatchProducer<RecordAggregator>>,
+}
+
+impl Client {
+    /// Creates a Client from the raw client.
+    pub(crate) fn new(raw_client: Arc<PartitionClient>, config: &KafkaConfig) -> Self {
+        let record_aggregator = RecordAggregator::new(config.max_batch_size.as_bytes() as usize);
+        let batch_producer = BatchProducerBuilder::new(raw_client.clone())
+            .with_compression(config.compression)
+            .with_linger(config.linger)
+            .build(record_aggregator);
+
+        Self {
+            raw_client,
+            producer: Arc::new(batch_producer),
+        }
+    }
+}
+
+/// Manages client construction and accesses.
+#[derive(Debug)]
+pub(crate) struct ClientManager {
+    config: KafkaConfig,
+    /// Top-level client in rskafka. All clients are constructed by this client.
+    client_factory: RsKafkaClient,
+    /// A pool maintaining a collection of clients.
+    /// Key: a topic. Value: the associated client of the topic.
+    client_pool: DashMap<Topic, ClientRef>,
+}
+
+impl ClientManager {
+    /// Tries to create a ClientManager.
+    pub(crate) async fn try_new(config: &KafkaConfig) -> Result<Self> {
+        // Sets backoff config for the top-level rskafka client and all clients constructed by it.
+        let backoff_config = BackoffConfig {
+            init_backoff: Duration::from_millis(500),
+            max_backoff: Duration::from_secs(10),
+            base: 2.,
+            // Stop reconnecting if the total wait time reaches the deadline.
+            deadline: Some(Duration::from_secs(60 * 5)),
+        };
```

Comment:
The same suggestion as the previous PR. I'd like to make sure this config can be configured by the users.

Reply:
It's now configurable.